src/test/CharterManager-unit.t.sol

```diff
@@ -0,0 +1,644 @@
+        (uint256 ink, uint256 art) = a.urn();
+        assertEq(ink, 100 * 1e18);
+        assertEq(art, 50 * 1e18);
+        assertEq(a.dai(), 50 * 1e45);
+        assertEq(a.gems(), 0);
+        assertEq(vat.dai(address(vow)), 0);
+        a.frob(-100 * 1e18, -50 * 1e18);
+        (ink, art) = a.urn();
+        assertEq(ink, 0);
+        assertEq(art, 0);
+        assertEq(a.dai(), 0);
+        assertEq(a.gems(), 100 * 1e18);
+    }
+
+    function test_frob_gate_nib() public {
+        (Usr a,) = init_user();
+        init_ilk_gate(address(a), 2 * NIB_ONE_PCT, 50 * 1e45);
+
+        a.join(100 * 1e6);
+        a.frob(100 * 1e18, 50 * 1e18);
+        (uint256 ink, uint256 art) = a.urn();
+        assertEq(ink, 100 * 1e18);
+        assertEq(art, 50 * 1e18);
+        assertEq(a.dai(), 49 * 1e45);
+        assertEq(a.gems(), 0);
+        assertEq(vat.dai(address(vow)), 1 * 1e45);
+
+        // frob out some of the funds
+        a.frob(-90 * 1e18, -49 * 1e18);
+        (ink, art) = a.urn();
+        assertEq(ink, 10 * 1e18);
+        assertEq(art, 1 * 1e18);
+        assertEq(a.dai(), 0);
+        assertEq(a.gems(), 90 * 1e18);
+
+        // force extra dai balance to cover for paid origination fee
+        cheat_get_dai(address(this), uint256(1 * 1e18));
+        dai.approve(address(daiJoin), 1 * 1e18);
+        daiJoin.join(address(a), 1 * 1e18);
+        assertEq(a.dai(), 1 * 1e45);
+
+        // repay remaining debt
+        a.frob(-10 * 1e18, -1 * 1e18);
+        (ink, art) = a.urn();
+        assertEq(ink, 0);
+        assertEq(art, 0);
+        assertEq(a.dai(), 0);
+        assertEq(a.gems(), 100 * 1e18);
+    }
+
+    function testFail_frob_undelegated_manager() public {
+        (Usr a,) = init_user();
+        init_ilk_gate(address(a), 0, 50 * 1e45);
+        a.nope(address(vat), address(manager));
+
+        a.join(100 * 1e6);
+        a.frob(100 * 1e18, 50 * 1e18);
+
+        // loan repayment should fail as its destination did not delegate the manager
+        a.frob(-100 * 1e18, -50 * 1e18);
+    }
+
+    function testFail_frob_gate_uline_exceeded() public {
+        (Usr a,) = init_user();
+        init_ilk_gate(address(a), 0, 50 * 1e45);
+
+        a.join(100 * 1e6);
+        a.frob(100 * 1e18, 60 * 1e18);
+    }
+
+    function testFail_frob_gate_other() public {
+        (Usr a, Usr b) = init_user();
+        init_ilk_gate(address(a), 0, 50 * 1e45);
+
+        b.join(100 * 1e6);
+        b.frob(100 * 1e18, 50 * 1e18);
+    }
+
+    function test_drip_withdraw() public {
+        init_ilk_ungate(10 * NIB_ONE_PCT);
+        (Usr a,) = init_user();
+
+        a.join(20 * 1e6);
+        a.frob(20 * 1e18, 20 * 1e18);
+        (uint256 ink, uint256 art) = a.urn();
+        assertEq(ink, 20 * 1e18);
+        assertEq(art, 20 * 1e18);
+        assertEq(a.dai(), 18 * 1e45);
+        assertEq(a.gems(), 0);
+        assertEq(vat.dai(address(vow)), 2 * 1e45);
+
+        hevm.warp(now + 365 days);
+        jug.drip(ilk);
```

Comment:
I strongly prefer the pattern of testing fee accumulation by calling `Vat.fold` directly, which removes the need to set up the Jug. We only care that `Ilk.rate != RAY` for this test; we should take the shortest route to get there.

Reply:
Done.